_SPIKE_.md

```diff
@@ -0,0 +1,527 @@
+the parameter via DataSource settings.
+
+```ts
+/**
+ * Description of capabilities offered by the connector backing the given
+ * datasource.
+ */
+export interface ConnectorCapabilities {
+  /**
+   * Maximum number of items allowed for `inq` operators.
+   * This value is used to split queries when resolving related models
+   * for a large number of source instances.
+   */
+  inqLimit?: number;
+}
+```
+
+To preserve backwards compatibility with existing repository implementation, we
+cannot add `ConnectorCapabilities` directly to the `Repository` class. We need
+to introduce a new interface instead that Repositories can (or may not)
+implement.
+
+```ts
+export interface WithCapabilities {
+  capabilities: ConnectorCapabilities;
+}
+```
+
+### MongoDB and `ObjectID` type
+
+MongoDB is tricky.
+
+- It uses a custom `ObjectID` type for primary keys.
+- `ObjectID` is represented as a `string` when converted to JSON
+- In queries, string values must be case to ObjectID, otherwise they are not
+  considered as the same value: `'some-id' !== ObjectID('some-id')`.
+
+As a result, both PK and FK properties must use `ObjectID` as the type, and
+coercion must be applied where necessary.
+
+Ideally, I'd like LB4 to define MongoDB PK and FKs as follows:
+
+- `{type: 'string', mongodb: {dataType: 'ObjectID'}}`
+
+Even better, `dataType: 'ObjectID'` should be automatically applied by the
+connector for PK and FKs referencing ObjectID PKs.
+
+For example:
+
+```ts
+@model()
+class Product {
+  @property({
+    type: 'string',
+    generated: true,
+    // ^^ implies dataType: 'ObjectID'
+  })
+  id: string;
+
+  @property({
+    type: 'string',
+    references: {
+      model: () => Category,
+      property: 'id',
+    },
+    // ^^ implies dataType: 'ObjectID' when Category is attached to MongoDB
+  })
+  categoryId: string;
+}
+```
+
+For v1, I suppose we can ask developers to provide dataType manually.
+
+```ts
+@model()
+class Product {
+  @property({
+    type: 'string',
+    generated: true,
+    mongodb: {dataType: 'ObjectID'},
+  })
+  id: string;
+
+  @property({
+    type: 'string',
+    mongodb: {dataType: 'ObjectID'},
+  })
+  categoryId: string;
+}
+```
+
+With this setup in place, `id` and `categoryId` properties should be always
+returned as strings from DAO and connector methods.
+
+This is tricky to achieve for the PK property, because juggler overrides
+user-supplied type with connector's default type when the PK is generated by the
+database. See
+[`DataSource.prototype.setupDataAccess()`](https://github.com/strongloop/loopback-datasource-juggler/blob/0c2bb81dace3592ecde8b9eccbd70d589da44d7d/lib/datasource.js#L713-L719)
+
+Can we rework MongoDB connector to hide ObjectID complexity as an internal
+implementation detail and always use string values in public APIs? Accept ids as
+strings and internally convert them to ObjectID. Convert values returned by the
+database from ObjectID to strings.
```

Comment:
@bajtos  If I am interpreting this correctly, you are suggesting to get rid of the `ObjectID` type altogether from the MongoDB connector. Right?

Reply:
I am proposing to modify MongoDB connector to hide `ObjectID` type as an internal implementation detail, an use `string` type in all other LoopBack layers (Models, Repositories, Controllers).

In the database, primary and foreign keys must be still stored as `ObjectID`, this is important e.g. for performance reasons. The connector will still have to deal with `ObjectID` type.

I just don't want `ObjectID` to leak outside of the connector.

BTW this will be a breaking change that will also break compatibility with LB3. I think it is not an issue, just want to be clear about the ramifications.

Also, these changes are not strictly required for the implementation of inclusion. It's a wider problem relevant to relations in general.

Reply:
> I just don't want ObjectID to leak outside of the connector.

Makes sense. There is an existing issue (I encountered while writing tests, but we decided to defer tackling it), in which a property typed as `ObjectID` is returned as string by the MongoDB driver. The proposed change will eliminate such issues.